services/ux-backend/handlers/onboardingtokens/handler.go

```diff
@@ -33,32 +36,44 @@ func HandleMessage(w http.ResponseWriter, r *http.Request, tokenLifetimeInHours
 
 func handlePost(w http.ResponseWriter, r *http.Request, tokenLifetimeInHours int) {
 	var storageQuotaInGiB *uint
+	// to ensure backward compatibility, if the role is empty, we assume it to be client
+	role := services.ClientRole
 	// When ContentLength is 0 that means request body is empty and
 	// storage quota is unlimited
 	var err error
 	if r.ContentLength != 0 {
-		var quota = struct {
-			Value uint   `json:"value"`
-			Unit  string `json:"unit"`
+		var body = struct {
+			SubjectRole string `json:"subjectRole"`
+			Value       uint   `json:"value"`
+			Unit        string `json:"unit"`
 		}{}
-		if err = json.NewDecoder(r.Body).Decode(&quota); err != nil {
+		if err = json.NewDecoder(r.Body).Decode(&body); err != nil {
 			http.Error(w, err.Error(), http.StatusBadRequest)
 			return
 		}
-
-		if quota.Value == 0 || quota.Value > math.MaxInt {
-			http.Error(w, fmt.Sprintf("invalid value sent in request body, value should be greater than 0 and less than %v: %v", math.MaxInt, quota.Value), http.StatusBadRequest)
+		if strings.ToLower(body.SubjectRole) == string(services.PeerRole) {
+			role = services.PeerRole
+		} else if strings.ToLower(body.SubjectRole) == string(services.ClientRole) || body.SubjectRole == "" {
+			// to ensure backward compatibility, if the role is empty, we assume it to be client
+			role = services.ClientRole
+			if body.Value == 0 || body.Value > math.MaxInt {
+				http.Error(w, fmt.Sprintf("invalid value sent in request body, value should be greater than 0 and less than %v: %v", math.MaxInt, body.Value), http.StatusBadRequest)
+				return
+			}
+			unitAsGiB, ok := unitToGib[body.Unit]
+			if !ok {
+				http.Error(w, fmt.Sprintf("invalid Unit type sent in request body, Valid types are [Gi,Ti,Pi]: %v", body.Unit), http.StatusBadRequest)
+				return
+			}
+			storageQuotaInGiB = ptr.To(unitAsGiB * body.Value)
+		} else {
+			http.Error(w, fmt.Sprintf("invalid OnboardingSubjectRole sent in request body, Valid roles are [mirroring-peer,client]: %v", body.SubjectRole), http.StatusBadRequest)
```

Comment:
The flow is structured incorrectly. The current structure allows sending peer role with quota fields and the quota fields are just ignored. This is the wrong behavior. I would expect a bad request response instead.

I would suggest using a switch case based on the role and delegate each of the 2 cases into separate functions. In addition, you could use the "default" case to return an error for an unsupported  role